backend/src/Database.hs

```diff
@@ -5,22 +5,31 @@ import Data.Pool (withResource)
 import Database.SQLite.Simple (Only (..), execute_, query, query_)
 import Database.SQLite.Simple qualified as SQL
 import Types.App (Env (..), log)
-import Types.DataField (PlayerId, PlayerName)
-import Types.Database (ReadPlayer (..), ReadProcessedGameReport (..), WritePlayer (..), WriteProcessedGameReport (..))
+import Types.DataField (PlayerId, PlayerName, Rating, Side)
+import Types.Database
+  ( ReadPlayer (..),
+    ReadProcessedGameReport (..),
+    WritePlayer (..),
+    WriteProcessedGameReport (..),
+    WriteRatingChange (..),
+  )
 
 data UnexpectedResultException = UnexpectedResultException deriving (Show)
 
 instance Exception UnexpectedResultException
 
-readSingle :: IO [Only r] -> IO r
-readSingle rs = rs >>= \case [Only r] -> pure r; _ -> throwIO UnexpectedResultException
+readZeroOrOne :: IO [r] -> IO (Maybe r)
+readZeroOrOne rs = rs >>= \case [] -> pure Nothing; [r] -> pure $ Just r; _ -> throwIO UnexpectedResultException
+
+readOne :: IO [r] -> IO r
+readOne rs = rs >>= \case [r] -> pure r; _ -> throwIO UnexpectedResultException
 
 initializeDatabase :: ReaderT Env IO ()
 initializeDatabase = do
   env <- ask
   liftIO . withResource env.dbPool $ \conn -> do
     log env.logger "Initializing database"
-    tableCount <- readSingle $ query_ conn "SELECT COUNT(*) FROM sqlite_master WHERE type='table';" :: IO Int
+    tableCount <- fromOnly <$> readOne (query_ conn "SELECT COUNT(*) FROM sqlite_master WHERE type='table';") :: IO Int
```

Comment:
`fromOnly`, cool!

Reply:
This is from the sqlite-simple library, and is a very common pattern in haskell. Since we often have values wrapped in some kind of container type (`Only` in this case, but we do it all the time) -- there's two main ways to "extract" the internal value. One is pattern matching, but the other is to provide a utility function that extracts the value for you.

Reply:
Also, not sure if it's obvious here, but the fmap is to apply the fromOnly inside the IO returned by readOne

Reply:
Still takes my brain an extra step to read `<$>` as `fmap` instead of thinking it's the applicative one `<*>` (idk, it's a mental block) so in that sense it was not 😂  but I did pick up on it being needed for lifting `fromOnly` into the IO! Baby steps